ext/standard/math.c

```diff
@@ -228,6 +228,32 @@ PHP_FUNCTION(abs)
 }
 /* }}} */
 
+/* {{{ Return the given value if in range of min and max */
+PHP_FUNCTION(clamp)
+{
+	zval *zvalue, *zmin, *zmax;
+
+	ZEND_PARSE_PARAMETERS_START(3, 3)
+		Z_PARAM_NUMBER(zvalue)
+		Z_PARAM_NUMBER(zmin)
+		Z_PARAM_NUMBER(zmax)
+	ZEND_PARSE_PARAMETERS_END();
+
+	if (zend_compare(zmin, zmax) > 0) {
+		zend_argument_value_error(2, "must be smaller than or equal to argument #3 ($max)");
+		RETURN_THROWS();
+	}
+
+	if (zend_compare(zmax, zvalue) == -1) {
+		RETURN_COPY(zmax);
+	} else if (zend_compare(zmin, zvalue) == -1) {
```

Comment:
If you do `if (zend_compare(zvalue, zmin) == -1) { return zmin } else { return zvalue }` that should handle the NAN value case, I think.

Reply:
> If you do `if (zend_compare(zvalue, zmin) == -1) { return zmin } else { return zvalue }` that should handle the NAN value case, I think.

That it did yes, changed which returns when `$num` and `$min` are equal to but no breaking change so. 👍

> By the way, you can use RETURN_COPY_VALUE here, as you are working on simple types only.

Sorry for asking but since there isn't much documentation for the PHP internals I have to ask. What is the difference between `RETURN_COPY` and `RETURN_COPY_VALUE`? From the name, it sounds like maybe `RETURN_COPY` would be better used for bigger structures like arrays and objects and `RETURN_COPY_VALUE` is better for simple types like ints, bools, and maybe strings?

Reply:
RETURN_COPY is copy the value and increment the refcount. RETURN_COPY_VALUE is just copy the value.

In this case, RETURN_COPY is semantically correct, but it so happens that you know that the copied value can only be an integer or float, so it will never have a refcount that could be incremented.